wgpu-hal/src/vulkan/instance.rs

```diff
@@ -1011,12 +1016,28 @@ impl crate::Surface for super::Surface {
             timeout_ns = u64::MAX;
         }
 
-        let wait_semaphore = sc.surface_semaphores[sc.next_surface_index];
+        let swapchain_semaphores_arc = swapchain.get_surface_semaphores();
+        // Nothing should be using this, so we don't block, but panic if we fail to lock.
+        let locked_swapchain_semaphores = swapchain_semaphores_arc
+            .try_lock()
+            .expect("Failed to lock a SwapchainSemaphores.");
+
+        // Wait for the previously acquired image used by the semaphore to be available.
+        swapchain.device.wait_for_fence(
+            fence,
+            locked_swapchain_semaphores.previously_used_submission_index,
+            timeout_ns,
+        )?;
```

Comment:
What does this actually accomplish? Aren't we going to wait for the `acquire` semaphore anyway?

Reply:
We must wait until the cpu timeline is sure that the semaphore has been signaled. Otherwise there will be two acquire operations that could signal the same semaphore.

Reply:
So, in other words:

If we were to remove this wait, then acquisition, command submission, and presentation would all set stuff in motion in the presentation engine or on the GPU, but they'd all return without blocking, so the CPU could just run the process full speed ahead, blow through the entire swapchain, come around to the front of the ring buffer again, and make a fresh call to `vkAcquireNextImage` passing the same semaphore as last time. This wait is the only thing that rate-limits that process to match actual presentation speed.

Is that correct?

Reply:
I've pushed another commit documenting this. If my explanation is correct, please mark this as resolved.